### ISSUE TITLE
GS1 QR code

### ISSUE THREAD

**Issue by dannyto** [soldair/node-qrcode#45]
Hi there!

Can we create gs1 qr code.

Thankz.

**Comment by soldair** [comment]
I'm not sure.
It seems like qrcode became a "gs1 standard" so i bet that this library can. There are constraints as to the dimensions of the code and other things hinted in the text here:

http://helpdesk.gs1.org/ArticlesBySubject.aspx?GS1%20QR%20Code&id=83706133-2072-e211-ad68-00155d644635

if someone would do a little research it would be welcome.

**Comment by dannyto** [comment]
Thank you, I hope so! 
Anyone can help please!

**Comment by soldair** [comment]
looks like this project implments them and has info on what it means to be a gs1 qrcode
https://github.com/bwipp/postscriptbarcode/wiki/GS1-Application-Identifier-Standard-Format

**Comment by robotiko** [comment]
@dannyto , @soldair, 
GS1 QR codes are standard QR codes that contains text with GS1 formatted data. 

The GS1 data is basically a set of GS1 Application Identifiers.
The Application Identifiers (AI) are numeric values defined in the GS1 catalog (http://www.gs1.org/barcodes-epcrfid-id-keys/gs1-general-specifications)   followed by the value for that (AI). 
The mandatory value is the (01) Global Trade Item Number (GTIN) that identifies the product/ data owner. 

All this GS1 jargon is quite confusing, but basically this is a catalog of identifiers with predefined values. Somehow is a schema in flat format. 
By itself the GS1 QR code defines how to structure the data contained, but not what data should be defined.  (Sintaxis but not semantics).
There are some fields and business scenarios  that have defined GS1 standards with flows and data  defined to be followed (GS1 guides that define the AI sets).

The main example provided in the https://github.com/bwipp/postscriptbarcode  that can be found http://www.gs1.org/docs/barcodes/GS1_Barcodes_Fact_Sheet-GS1_2D_symbols.pdf 
Defines  an example for the  "extended packaging" scenario that in "Retail world" means additional online information for a product package. 
It simply requires a global id (01) associated to the product and the Extended Packaging URL identifier (8200) .

In conclusion, almost any QR generator can work with GS1 QR codes, but it is too open. 
For a mode detailed and constrained content, it has to be application specific. 

So Yes, this can create a GS1 QR code. 
About constraints about to the dimensions of the code,  GS1 usually define recommendations, not restrictions because each case is different and at end it will depend on real world usage. It is not the same to print a barcode on a makeup eye pen (really tiny) than on big  cereal box. :)  

Do you have any specific usage in mind?

**Comment by soldair** [comment]
wow thanks @robotiko for such a detailed response!
i just wanted to learn a bit and make sure that this library is at least fulfilling the base needs of it's users respecting it's scope.

@dannyto i hope this helps. The answer seems like yes but you have to add the gs1 fields to the data you put in the qrcode.

**Comment by robotiko** [comment]
@soldair  I hope it helps.

Any QR code will work with GS1 codes and the correctness is in the code info, not the barcode itself that is a mere container.

Deal with GS1 coding requires a lot of GS1 knowledge  about the standards, guides, etc. 
Even the basic & required GTIN requires to join GS1 to be a member and get the unique identifier, so go deeper into the standards is a must to be able use them.

**Comment by dannyto** [comment]
Thanks all you guys!
Those info are so useful.

**Comment by soldair** [comment]
closing but ill link to this issue from the readme about GS1

**Comment by fanke** [comment]
dear soldair, the "node-qrcode" is greate, but when can you support "i would like to switch to a js only image encoder to remove these deps", what i think will be greater, thank you!

**Comment by soldair** [comment]
i think so too. feel free to dig around for a god one I'm always accepting pull requests.

**Comment by jsweber** [comment]
how to set canvas'size. such as width and height

**Comment by ch250001757** [comment]
can add a logo in the middle

**Comment by cll123456** [comment]
can add a logo in the middle

**Comment by baobinaqd** [comment]
if create qrcode is a img，I want to set some attributes，eg: alt,title, can I do ?

**Comment by m-hey** [comment]
why same opts, but second setting toCanvas Api not show qrcode？anyone has this question?

**Comment by Dhanraj09** [comment]
its Qr code is not working googlepay & paytm
